RileyLinkKitTests/PumpOpsSynchronousTests.swift

```diff
@@ -73,202 +82,433 @@ class PumpOpsSynchronousTests: XCTestCase {
         
-        assertArray(timestampedEvents, doesntContainPumpEvent: event2010)
+        XCTAssertFalse(hasMoreEvents)
     }
     
-    func testShouldContainEventWhenEstimatedTimeDeltaAllowanceBefore523AdjustedStartTime() {
-        let event2010 = createSquareBolusEvent2010()
+    func testSquareBolusDiscontinuityAfterDelayedAppendEventCancels() {
+        setUpTestWithPumpModel(.Model523)
+        
+        let squareBolus2016 = createSquareBolusEvent2016()
+        let events:[PumpEvent] = [createSquareBolusEvent2010(), squareBolus2016]
+        let (_, _, cancelled) = sut.convertPumpEventToTimestampedEvents(pumpEvents: events, startDate: Date.distantPast, pumpModel: pumpModel)
+        
+        XCTAssertTrue(cancelled)
+    }
 
-        let (timestampedEvents, _) = runDeltaAllowanceTimeTest(pumpEvent: event2010, timeIntervalAdjustment: TimeInterval(minutes:9))
+// MARK: Final Regular Bolus Discontinuity Event (after Square Bolus)
+    func testFinalRegularDiscontinuityBolusDoesReturnEventFor523() {
+        setUpTestWithPumpModel(.Model523)
         
-        assertArray(timestampedEvents, containsPumpEvent: event2010)
+        let regularBolusEvent = createBolusEvent2011()
+        let events:[PumpEvent] = [createSquareBolusEvent2010(), createSquareBolusEvent2010(), regularBolusEvent]
+        let (timeStampedEvents, _, _) = sut.convertPumpEventToTimestampedEvents(pumpEvents: events, startDate: Date.distantPast, pumpModel: pumpModel)
+        
+        //It should be returned
+        XCTAssertTrue(array(timeStampedEvents, containsPumpEvent: regularBolusEvent))
     }
     
-    func test522EstimatedTimeDeltaAllowanceBeforeAdjustedStartTime() {
-        loadTestWithPumpModel(.Model522)
+    func test523FinalRegularDiscontinuityHasMoreEvents() {
+        setUpTestWithPumpModel(.Model523)
+        
+        let regularBolusEvent = createBolusEvent2011()
+        let events:[PumpEvent] = [createSquareBolusEvent2010(), createSquareBolusEvent2010(), regularBolusEvent]
+        let (_, hasMoreEvents, _) = sut.convertPumpEventToTimestampedEvents(pumpEvents: events, startDate: Date.distantPast, pumpModel: pumpModel)
         
+        XCTAssertTrue(hasMoreEvents)
+    }
+    
+    func test523FinalRegularDiscontinuityDoesntCancelsOperation() {
+        setUpTestWithPumpModel(.Model523)
+        
+        let regularBolusEvent = createBolusEvent2011()
+        let events:[PumpEvent] = [createSquareBolusEvent2010(), createSquareBolusEvent2010(), regularBolusEvent]
+        let (_, _, cancelledEarly) = sut.convertPumpEventToTimestampedEvents(pumpEvents: events, startDate: Date.distantPast, pumpModel: pumpModel)
+        
+        // this doesn't trigger the out of order event cancellation
+        XCTAssertFalse(cancelledEarly)
+    }
+    
+// MARK: Square Bolus Event before starttime (offset 9 minutes)
+    func testSquareBolusEventBeforeStartTimeDoesntContainEvent() {
```

Comment:
The fact that sometimes square boluses before start date are returned (523) and sometimes they aren't (522), is a side effect of our deciding to skip delayed append events for time cutoff checking.  It's *not* something that we want to assert, as it is not really desired behavior, but it's acceptable behavior.